pkg/merkletree/tree.go

```diff
@@ -0,0 +1,173 @@
+package merkletree
+
+import (
+	"bytes"
+	"errors"
+	"sort"
+
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// DefaultHashFunc is the default hash function in case none is passed.
+var DefaultHashFunc = crypto.Keccak256
+
+// MerkleTree is a binary Merkle Tree implemenation.
+type MerkleTree struct {
+	root   *Node
+	leaves []*Node
+
+	hashFunc func(...[]byte) []byte
+}
+
+// Node represents a Node of MerkleTree.
+type Node struct {
+	parent, left, right *Node
+	hash                []byte
+}
+
+func (n *Node) isLeaf() bool {
+	return n.left == nil && n.right == nil
+}
+
+// NewTree builds a new Merkle Tree.
+func NewTree(leaves [][]byte, hashFunc func(...[]byte) []byte) (*MerkleTree, error) {
```

Comment:
You can pass the hash function. It helps with testing and gives clients the freedom to choose. I picked the signature `func(...[]byte) []byte` because the `hash.Hash` interface is quite cumbersome. But maybe will change this in the future because in some situation the size of the hash is needed, and this signature does not provided that.